### ISSUE TITLE
The typechecker really hates the spaceship operator

### ISSUE THREAD

**Issue by doublecompile** [facebook/hhvm#7037]
While HHVM seems to execute the code below correctly, the typechecker doesn't seem to be _hip_ enough to recognize the spaceship operator (AKA the Combined Comparison Operator, the Three-Way Comparison Operator). I'm guessing it's interpreting it as a less-than-or-equals.
### HHVM Version

3.13.1
#### SpaceshipOne.hh

``` hack
<?hh // strict
class SpaceshipOne
{
    public function compare(string $a, string $b): int
    {
        return $a <=> $b;
    }
}
```
#### SpaceshipTwo.hh

``` hack
<?hh // strict
class SpaceshipTwo
{
    public function sortLambda(Vector<string> $v): void
    {
        usort($v, ($a, $b) ==> $a <=> $b);
    }
}
```
#### SpaceshipThree.hh

``` hack
<?hh // strict
class SpaceshipThree
{
    public function sortClosure(Vector<string> $v): void
    {
        usort($v, function ($a, $b) {
            return $a <=> $b;
        });
    }
}
```
### Actual result

Running `hh_client` gives:

```
SpaceshipOne.hh:7:16,24: Invalid return type (Typing[4110])
  SpaceshipOne.hh:5:52,54: This is an int
  SpaceshipOne.hh:7:16,24: It is incompatible with a bool because this is the result of a comparison
SpaceshipOne.hh:7:16,24: Typing error (Typing[4110])
  SpaceshipOne.hh:5:29,34: This is a string
  SpaceshipOne.hh:7:21,24: It is incompatible with a bool because this is the result of a comparison
SpaceshipOne.hh:7:21,21: Expected expression (Parsing[1002])
SpaceshipOne.hh:7:21,24: Typing error (Typing[4110])
  SpaceshipOne.hh:7:21,21: This is a nullable type
  SpaceshipOne.hh:5:40,45: It is incompatible with a string
SpaceshipThree.hh:8:25,25: Expected expression (Parsing[1002])
SpaceshipTwo.hh:7:37,37: Expected expression (Parsing[1002])
```

Meanwhile the code itself works as expected.

``` hack
$v1 = ImmVector{'alpha', 'charlie', 'bravo', 'echo', 'delta'};

$s1 = new SpaceshipOne();
$c = $s1->compare('foo', 'bar');
echo "$c ", gettype($c), PHP_EOL;
// prints 1 integer

$v2 = new Vector($v1);
$s2 = new SpaceshipTwo();
$s2->sortLambda($v2);
print_r($v2);
// prints HH\Vector Object ( [0] => alpha [1] => bravo [2] => charlie [3] => delta [4] => echo )

$v3 = new Vector($v1);
$s3 = new SpaceshipThree();
$s3->sortClosure($v3);
print_r($v3);
// prints HH\Vector Object ( [0] => alpha [1] => bravo [2] => charlie [3] => delta [4] => echo )
```

**Comment by usox** [comment]
@orvid @doublecompile 
Seems to be fixed with 3.21

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

Thanks for the repros.
Hack understand `<=>` correctly in 4.56 and according to usox also in 3.21.
Hack models the expression as returning an int and accepts the expression only if both the LHS and the RHS are **both**:
 - num (or a part of the union, so `float <=> int` is allowed)
 - string
 - Datetime or DatetimeImmutable
